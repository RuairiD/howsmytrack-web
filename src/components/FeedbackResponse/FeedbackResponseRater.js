import React, { useState, useMemo } from "react";
import { useMutation } from "react-query";
import axios from "axios";

import { Button, Rate } from "antd";
import { Span } from "lemon-reset";
import apiRoot from "../../apiRoot";

const RATE_FEEDBACK_RESPONSE_MUTATION = `mutation RateFeedbackResponse($feedbackResponseId: Int!, $rating: Int!) {
    rateFeedbackResponse(feedbackResponseId: $feedbackResponseId, rating: $rating) {
        success
        error
    }
}`;

const RATING_TOOLTIP_TEXTS = [
    "This feedback is irrelevant and completely unhelpful.",
    "This feedback is vague, unhelpful, irrelevant or poorly written.",
    "This feedback is useful.",
    "This feedback is relevant, helpful and thoughtful.",
    "This feedback is deeply thoughtful, well-written and very constructive.",
];

const FeedbackResponseRater = ({
    feedbackResponseId,
    currentRating,
}) => {
    const [rating, setRating] = useState(currentRating);

    const onRatingChange = (newRating) => {
        setRating(newRating);
    };

    const submitRating = () => (
        axios.post(`${apiRoot()}/graphql/`, {
            query: RATE_FEEDBACK_RESPONSE_MUTATION,
            variables: {
                feedbackResponseId,
                rating,
            },
        }).then((response) => response.data.data.rateFeedbackResponse)
    );

    const [submitRatingMutate, { data, isLoading }] = useMutation(submitRating);

    const submitted = useMemo(() => (!!currentRating || (data && data.success)), [currentRating, data]);

    return (
        <Span style={{ float: "left", paddingBottom: "0.25em" }}>
            <Rate
                style={{
                    color: "#000000",
                    marginRight: "1em",
                }}
                allowClear
                tooltips={RATING_TOOLTIP_TEXTS}
                value={rating}
                disabled={submitted || isLoading}
                onChange={onRatingChange}
            />
            <Button
                type="primary"
                loading={isLoading}
                disabled={submitted || isLoading || !rating}
                onClick={submitRatingMutate}
            >
                {submitted && "Rated"}
                {!submitted && "Rate"}
            </Button>
        </Span>
    );
};

export default FeedbackResponseRater;
