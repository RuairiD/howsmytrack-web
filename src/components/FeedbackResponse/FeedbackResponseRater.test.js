import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { act } from "react-dom/test-utils";
import waitForExpect from "wait-for-expect";
import axios from "axios";
import FeedbackResponseRater from "./FeedbackResponseRater";

jest.mock("axios");

describe("FeedbackResponseRater", () => {
    it("renders an unset Rate and a disabled submission Button if no currentRating is provided", () => {
        const wrapper = shallow(
            <FeedbackResponseRater
                feedbackResponseId={1901}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders a Rate preset with the currentRating and a disabled submission Button since the feedback has already been rated", () => {
        const wrapper = shallow(
            <FeedbackResponseRater
                feedbackResponseId={1901}
                currentRating={5}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("submits a rating and disables the form if successful", async () => {
        axios.post.mockResolvedValue({
            data: {
                data: {
                    rateFeedbackResponse: {
                        success: true,
                        error: null,
                    },
                },
            },
        });

        const wrapper = mount(
            <FeedbackResponseRater
                feedbackResponseId={1901}
            />,
        );

        act(() => wrapper.find("Rate").get(0).props.onChange(5));
        await act(() => wrapper.find("Button").get(0).props.onClick());

        await waitForExpect(async () => {
            wrapper.update();
            expect(wrapper.find("Button").get(0).props.disabled).toBeTruthy();
            expect(wrapper.find("Rate").get(0).props.disabled).toBeTruthy();
            expect(wrapper.text().includes("Rated")).toBeTruthy();
        });
    });

    it("submits a rating but does not disable the form if there is an error", async () => {
        axios.post.mockResolvedValue({
            data: {
                data: {
                    rateFeedbackResponse: {
                        success: false,
                        error: "error",
                    },
                },
            },
        });

        const wrapper = mount(
            <FeedbackResponseRater
                feedbackResponseId={1901}
            />,
        );

        act(() => wrapper.find("Rate").get(0).props.onChange(5));
        await act(async () => wrapper.find("Button").get(0).props.onClick());

        await waitForExpect(async () => {
            wrapper.update();
            expect(wrapper.find("Rate").get(0).props.disabled).not.toBeTruthy();
            expect(wrapper.find("Button").get(0).props.disabled).not.toBeTruthy();
            expect(wrapper.text().includes("Rated")).not.toBeTruthy();
        });
    });
});
