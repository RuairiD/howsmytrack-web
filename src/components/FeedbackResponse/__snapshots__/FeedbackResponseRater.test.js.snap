// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackResponseRater renders a Rate preset with the currentRating and a disabled submission Button since the feedback has already been rated 1`] = `
<Span
  className=""
  style={
    Object {
      "float": "left",
      "paddingBottom": "0.25em",
    }
  }
>
  <Rate
    allowClear={true}
    character={
      <Icon
        theme="filled"
        type="star"
      />
    }
    disabled={true}
    onChange={[Function]}
    style={
      Object {
        "color": "#000000",
        "marginRight": "1em",
      }
    }
    tooltips={
      Array [
        "This feedback is irrelevant and completely unhelpful.",
        "This feedback is vague, unhelpful, irrelevant or poorly written.",
        "This feedback is useful.",
        "This feedback is relevant, helpful and thoughtful.",
        "This feedback is deeply thoughtful, well-written and very constructive.",
      ]
    }
    value={5}
  />
  <Button
    block={false}
    disabled={true}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[Function]}
    type="primary"
  >
    Rated
  </Button>
</Span>
`;

exports[`FeedbackResponseRater renders an unset Rate and a disabled submission Button if no currentRating is provided 1`] = `
<Span
  className=""
  style={
    Object {
      "float": "left",
      "paddingBottom": "0.25em",
    }
  }
>
  <Rate
    allowClear={true}
    character={
      <Icon
        theme="filled"
        type="star"
      />
    }
    disabled={false}
    onChange={[Function]}
    style={
      Object {
        "color": "#000000",
        "marginRight": "1em",
      }
    }
    tooltips={
      Array [
        "This feedback is irrelevant and completely unhelpful.",
        "This feedback is vague, unhelpful, irrelevant or poorly written.",
        "This feedback is useful.",
        "This feedback is relevant, helpful and thoughtful.",
        "This feedback is deeply thoughtful, well-written and very constructive.",
      ]
    }
  />
  <Button
    block={false}
    disabled={true}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[Function]}
    type="primary"
  >
    Rate
  </Button>
</Span>
`;
